python/ray/serve/batching.py

```diff
@@ -172,45 +190,82 @@ async def _consume_func_generator(
         FINISHED_TOKEN = None
 
         try:
-            futures = initial_futures
+            futures = deque(initial_futures)
+            assert len(futures) == input_batch_length
+
             async for results in func_generator:
                 self._validate_results(results, input_batch_length)
-                next_futures = []
-                for result, future in zip(results, futures):
+                for idx in range(input_batch_length):
+                    result, future = results[idx], futures[0]
+
                     if future is FINISHED_TOKEN:
                         # This caller has already terminated.
-                        next_futures.append(FINISHED_TOKEN)
-                    elif result in [StopIteration, StopAsyncIteration]:
+                        futures.append(FINISHED_TOKEN)
+                    elif result in USER_CODE_STREAMING_SENTINELS:
                         # User's code returned sentinel. No values left
                         # for caller. Terminate iteration for caller.
-                        future.set_exception(StopAsyncIteration)
-                        next_futures.append(FINISHED_TOKEN)
+                        _set_exception_if_not_done(future, StopAsyncIteration)
+                        futures.append(FINISHED_TOKEN)
                     else:
                         next_future = get_or_create_event_loop().create_future()
-                        # If the client has disconnected, the future is canceled.
-                        # We should only set result when the connection is still live.
-                        if not future.cancelled():
-                            future.set_result(_GeneratorResult(result, next_future))
-                        next_futures.append(next_future)
-                futures = next_futures
+                        _set_result_if_not_done(
+                            future, _GeneratorResult(result, next_future)
+                        )
+                        futures.append(next_future)
+
+                    # Remove processed future.
+                    futures.popleft()
```

Comment:
We'd dequeue first, then handle it

Reply:
Synced offline. We'll continue to dequeue it last, and I'll add a comment explaining why.

Reply:
I added a comment.